CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/SyntaxNodeFile.swift

```diff
@@ -0,0 +1,494 @@
+      StmtSyntax("return nil")
+    }
+
+    try! VariableDeclSyntax("public static var structure: SyntaxNodeStructure") {
+      let choices = ArrayExprSyntax {
+        for choice in choices {
+          ArrayElementSyntax(
+            expression: ExprSyntax(".node(\(raw: choice.typeName).self)")
+          )
+        }
+      }
+
+      StmtSyntax("return .choices(\(choices))")
+    }
+  }
+}
+
+fileprivate extension Node {
+  func generateInitializerDeclHeader(optionalBaseAsMissing: Bool) -> PartialSyntaxNodeString {
+    if children.isEmpty {
+      return "public init()"
+    }
+
+    var genericParamNames: [String: Int] = [:]
+    var genericParams: [String] = []
+
+    func createFunctionParameterSyntax(for child: Child) -> FunctionParameterSyntax {
+      var paramType: String
+      if !child.kind.isNodeChoicesEmpty {
+        paramType = child.name
+      } else if child.hasBaseType {
+        if optionalBaseAsMissing {
+          paramType = "Missing\(child.typeName)"
+        } else {
+
+          // If we have a base type, make the initializer generic over its
+          // protocol instead.
+          let index = child.swiftName.index(child.swiftName.startIndex, offsetBy: 1)
+          paramType = child.swiftName[..<index].uppercased()
+
+          let paramCount = (genericParamNames[paramType] ?? 0) + 1
+          genericParamNames[paramType] = paramCount
+
+          if paramCount > 1 {
+            paramType += "\(paramCount)"
+          }
+          genericParams.append("\(paramType): \(child.typeName)Protocol")
+        }
+      } else {
+        paramType = child.typeName
+      }
+
+      if child.isOptional {
+        paramType += "?"
+      }
+
+      let label = child.isUnexpectedNodes ? "_ " : ""
+
+      return FunctionParameterSyntax("\(label)\(child.swiftName): \(paramType)\(child.defaultInitialization)", for: .functionParameters)
+    }
+
+    let params = FunctionParameterListSyntax {
+      FunctionParameterSyntax("leadingTrivia: Trivia? = nil", for: .functionParameters)
+
+      for child in children {
+        createFunctionParameterSyntax(for: child)
+      }
+
+      FunctionParameterSyntax("trailingTrivia: Trivia? = nil", for: .functionParameters)
+    }
+
+    if genericParams.isEmpty {
+      return """
+        public init(
+        \(params)
+        )
+        """
+    } else {
+      let generics = GenericParameterClauseSyntax(
+        genericParameterList: GenericParameterListSyntax {
+          for param in genericParams {
+            GenericParameterSyntax(name: .identifier(param))
+          }
+        }
+      )
+
+      return """
+        public init\(generics)(
+        \(params)
+        )
+        """
+    }
+  }
+}
+
+fileprivate extension Child {
+  /// If the type has a default value, return an expression of the form
+  /// ` = default_value` that can be used as the default value to for a
+  /// function parameter. Otherwise, return an empty string.
+  var defaultInitialization: String {
```

Comment:
This could return `InitializerClauseSyntax?` instead of a string for a little more structure.